mypy/checker.py

```diff
@@ -2744,21 +2744,44 @@ def check_multiple_inheritance(self, typ: TypeInfo) -> None:
         if len(typ.bases) <= 1:
             # No multiple inheritance.
             return
+
         # Verify that inherited attributes are compatible.
-        mro = typ.mro[1:]
-        for i, base in enumerate(mro):
+        # Construct a "typed" MRO that follows regular MRO order, but includes instances
+        # parametrized with their generic args.
+        # This detects e.g. `class A(Mapping[int, str], Iterable[str])` correctly.
+        # For each MRO entry, include it parametrized according to each base inheriting
+        # from it.
+        typed_mro = [
+            map_instance_to_supertype(base, parent)
```

Comment:
I am still confusing by this line. Seems like we don't propogate real types to whole tree. We are using just a current class bases, so it propogates options to class+2 level maximum

```python
class A[T]: ... 
class B[T]: ...  # <- str propagated here only
class C(B[str]): ...
class D(C): ...
```

I am afraid, that **A** doesn't know about str in **D** class analyzing.
I think, we should build the tree recursively - it was implemented in my PR

Reply:
Does `map_instance_to_supertype(C(B[str]), A[t])` works?

Reply:
Case could be even harder:

```python
class A[T]: ... 
class B(A[str]): ...
class C(B): ...
class D(C): ...
```

Does `map_instance_to_supertype(C, A[T])` works as expected?

Reply:
In the first case `typed_mro` for `class Foo(D, int)` is 

```
[__main__.D, __main__.C, __main__.B[builtins.str], __main__.A[builtins.str], builtins.int, builtins.object, builtins.object]
```

In the second case it is (also for `class Foo(D, int)`):

```
[__main__.D, __main__.C, __main__.B, __main__.A[builtins.str], builtins.int, builtins.object, builtins.object]
```

And the following raises a dianostic as expected:

```python
from typing import Generic, TypeVar

T = TypeVar("T")
class A(Generic[T]):
    foo: T
class B(A[str]): ...
class C(B): ...
class D(C): ...
class Foo(D, A[T]): ...  # E: Definition of "foo" in base class "A" is incompatible with definition in base class "A"
```